Parse/src/main/java/com/parse/ParseObject.java

```diff
@@ -921,8 +939,34 @@ protected boolean visit(Object object) {
           builder.put(KEY_ACL, acl);
           continue;
         }
+        if (key.equals(KEY_SELECTED_KEYS)) {
+          JSONArray safeKeys = json.getJSONArray(key);
+          if (safeKeys.length() > 0) {
+            Collection<String> set = new HashSet<>();
+            for (int i = 0; i < safeKeys.length(); i++) {
+              // Don't add nested keys.
+              String safeKey = safeKeys.getString(i);
+              if (safeKey.contains(".")) safeKey = safeKey.split(".")[0];
+              set.add(safeKey);
+            }
+            builder.safeKeys(set);
+          }
+          continue;
+        }
 
         Object value = json.get(key);
+        if (value instanceof JSONObject && json.has(KEY_SELECTED_KEYS)) {
+          // This might be a ParseObject. Pass selected keys to understand if it is complete.
+          JSONArray selectedKeys = json.getJSONArray(KEY_SELECTED_KEYS);
+          JSONArray nestedKeys = new JSONArray();
+          for (int i = 0; i < selectedKeys.length(); i++) {
+            String nestedKey = selectedKeys.getString(i);
+            if (nestedKey.startsWith(key+".")) nestedKeys.put(nestedKey.substring(key.length()+1));
```

Comment:
At first (942-954) the current object will be given its own direct keys (without dots, e.g. b). 
Later (958+) we extract the child object subkeys (e.g. b.c.d.e). That is passed to decoder.decode(), which will pass back the JSON object to this method. It will get ‘b’ as a safe key, and will pass ‘c.d.e’... and so on. It goes this way with nested iterations, until there are no nested ParseObjects.